File: app/src/main/java/com/example/unsplash/view/fragments/ListFragment.java
package com.example.unsplash.view.fragments;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.transition.TransitionInflater;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import com.example.unsplash.R;
import com.example.unsplash.model.models.Collection;
import com.example.unsplash.model.models.MyLikeChangerObject;
import com.example.unsplash.view.MainActivity;
import com.example.unsplash.view.adapters.MyPagedListAdapter;
import com.example.unsplash.model.models.Photo;
import com.example.unsplash.view.adapters.PagedListOnClickListener;
import com.example.unsplash.viewmodel.PhotoViewModel;

import java.util.Objects;


public class ListFragment extends Fragment {
    MyPagedListAdapter mAdapter;
    RecyclerView rv;
    Toolbar toolbar;
    final int numberOfColumns = 2;
    PagedListOnClickListener listener;
    public PhotoViewModel photoViewModel;
    private SwipeRefreshLayout swipeContainer;
    ImageView favouritesButton;

    private void refreshList() {
        Objects.requireNonNull(photoViewModel.photoPagedList.getValue()).getDataSource().invalidate();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        photoViewModel = ViewModelProviders.of(Objects.requireNonNull(getActivity())).get(PhotoViewModel.class);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_list, container, false);
        ((MainActivity) Objects.requireNonNull(getActivity())).showNavBar();
        viewInit(view);
        photoViewModel.photoPagedList.observe(this, photos -> mAdapter.submitList(photos));
        photoViewModel.photoLikeChangerObject.observe(this, myLikeChangerObject -> {
            assert myLikeChangerObject != null;
            if (!(myLikeChangerObject.getPosition() == -1)) {
                if (mAdapter.getCurrentList() != null) {
                    Objects.requireNonNull(Objects.requireNonNull(mAdapter.getCurrentList()).get(myLikeChangerObject.getPosition())).setLikedByUser(myLikeChangerObject.isLiked());

                    //just increment or decrement likes value
                    if (myLikeChangerObject.isLiked()) {
                        Objects.requireNonNull(Objects.requireNonNull(mAdapter.getCurrentList()).get(myLikeChangerObject.getPosition())).setLikes(Objects.requireNonNull(Objects.requireNonNull(mAdapter.getCurrentList()).get(myLikeChangerObject.getPosition())).getLikes() + 1);
                    } else {
                        Objects.requireNonNull(Objects.requireNonNull(mAdapter.getCurrentList()).get(myLikeChangerObject.getPosition())).setLikes(Objects.requireNonNull(Objects.requireNonNull(mAdapter.getCurrentList()).get(myLikeChangerObject.getPosition())).getLikes() - 1);
                    }

                    mAdapter.notifyItemChanged(myLikeChangerObject.getPosition());
                    MyLikeChangerObject my = new MyLikeChangerObject("a", false, -1);
                    photoViewModel.changeLike(my);
                }
            }
        });

        return view;
    }

    private void viewInit(View view) {
        rv = view.findViewById(R.id.rView);
        swipeContainer = view.findViewById(R.id.swipeContainer);

        swipeContainer.setColorSchemeResources(android.R.color.holo_blue_bright,
                android.R.color.holo_green_light,
                android.R.color.holo_orange_light,
                android.R.color.holo_red_light);


        swipeContainer.setOnRefreshListener(() -> {
            swipeContainer.setRefreshing(true);
            refreshList();
            swipeContainer.setRefreshing(false);
        });

        // todo "hide nav bar in imagefragment and favouritefragment"
        rv.setLayoutManager(new GridLayoutManager(view.getContext(), numberOfColumns));
        listenerInit();
        mAdapter = new MyPagedListAdapter(getActivity(), listener);
        rv.setAdapter(mAdapter);
        toolbar = view.findViewById(R.id.toolbar);
        ((AppCompatActivity) Objects.requireNonNull(getActivity())).setSupportActionBar(toolbar);
        favouritesButton = view.findViewById(R.id.favourite);
        favouritesButton.setOnClickListener(v -> {
            FavouritesFragment favouritesFragment = new FavouritesFragment();
            FragmentManager manager = (Objects.requireNonNull(getActivity()))
                    .getSupportFragmentManager();
            assert manager != null;
            manager.beginTransaction()
                    .setReorderingAllowed(true)
                    .replace(R.id.container, favouritesFragment)
                    .addToBackStack(null)
                    .commit();
        });
    }

    private void listenerInit() {
        listener = new PagedListOnClickListener() {
            @Override
            public void onClick(View view, Photo photo, int position) {
                Bundle bundle = new Bundle();
                assert photo != null;
                bundle.putString("URI", photo.getUrls().getRegular());
                bundle.putInt("SMTH", photo.getLikes());
                bundle.putString("TRANS", view.getTransitionName());
                bundle.putString("ID", photo.getId());
                bundle.putBoolean("ISLIKED", photo.getLikedByUser());
                bundle.putInt("POS", position);
                setReenterTransition(TransitionInflater
                        .from(getContext()).inflateTransition(android.R.transition.move).setDuration(100));
                ImageFragment imageFragment = new ImageFragment();
                imageFragment.setArguments(bundle);
                FragmentManager manager = (Objects.requireNonNull(getActivity()))
                        .getSupportFragmentManager();
                assert manager != null;
                manager.beginTransaction()
                        .setReorderingAllowed(true)
                        // animation works well on emulator, but not on 21api device
                        .addSharedElement(view, view.getTransitionName())
                        .replace(R.id.container, imageFragment)
                        .addToBackStack(null)
                        .commit();
            }

            @Override
            public void onClickCollection(View view, Collection collection) {

            }
        };
    }

}

File: app/src/main/java/com/example/unsplash/view/fragments/ListFragment.kt
package com.example.unsplash.view.fragments

import android.app.Activity
import android.content.Intent
import android.content.Intent.FLAG_ACTIVITY_NEW_TASK
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.PopupMenu
import androidx.appcompat.widget.SearchView.OnQueryTextListener
import androidx.core.view.doOnPreDraw
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.paging.PagedList
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.GridLayoutManager
import androidx.transition.TransitionInflater
import com.example.unsplash.R
import com.example.unsplash.model.models.MyLikeChangerObject
import com.example.unsplash.model.models.Photo
import com.example.unsplash.view.MainActivity
import com.example.unsplash.view.adapters.MyPagedListAdapter
import com.example.unsplash.viewmodel.PhotoViewModel
import dagger.android.support.DaggerFragment
import kotlinx.android.synthetic.main.fragment_list.*
import javax.inject.Inject


class ListFragment : DaggerFragment() {
    private lateinit var mAdapter: MyPagedListAdapter
    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private val photoViewModel: PhotoViewModel by lazy {
        ViewModelProvider(this, viewModelFactory)[PhotoViewModel::class.java]
    }
    private var isSearching = false
    private val observer = Observer<PagedList<Photo>> { mAdapter.submitList(it) }

    private fun refreshList() {
        photoViewModel.list.value?.dataSource?.invalidate()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        exitTransition =
            TransitionInflater.from(context).inflateTransition(android.R.transition.fade)
                .setDuration(300)
        enterTransition =
            TransitionInflater.from(context).inflateTransition(android.R.transition.fade)
                .setDuration(300)
        addObservers()
        mAdapter = MyPagedListAdapter(
            requireContext(),
            photoClickListener = { itemView, photo, i -> photoClick(itemView, photo, i) })
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_list, container, false)
        postponeEnterTransition()
        (activity as MainActivity).showNavBar()
        return view
    }

    private fun addObservers() {
        photoViewModel.list.observe(this, observer)
        photoViewModel.photoLikeChangerObject.observe(this, Observer {
            if (it != null) {
                if (it.position != -1) {
                    if (mAdapter.currentList != null) {
                        mAdapter.currentList!![it.position]?.likedByUser = it.isLiked
                        // just increment or decrement likes value
                        if (it.isLiked) {
                            mAdapter.currentList!![it.position]?.likes =
                                mAdapter.currentList!![it.position]?.likes?.plus(1)
                        } else {
                            mAdapter.currentList!![it.position]?.likes =
                                mAdapter.currentList!![it.position]?.likes?.minus(1)
                        }
                        mAdapter.notifyItemChanged(it.position)
                        val my = MyLikeChangerObject("a", false, -1)
                        photoViewModel.changeLike(my)
                    }
                }
            }
        })
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewInit()
        (view.parent as? ViewGroup)?.doOnPreDraw {
            startPostponedEnterTransition()
        }
    }

    private fun viewInit() {
        (activity as MainActivity).showFab()
        list_rv.layoutManager = GridLayoutManager(requireContext(), 2)
        list_rv.adapter = mAdapter
        list_rv.setEmptyView(empty)
        (activity as AppCompatActivity).setSupportActionBar(list_toolbar)
        favourite.setOnClickListener {
            val favouritesFragment = FavouritesFragment()
            activity?.supportFragmentManager?.beginTransaction()
                ?.replace(R.id.container, favouritesFragment)
                ?.addToBackStack(null)
                ?.commit()
        }
        list_swipe_container.setOnRefreshListener {
            if (!isSearching) {
                list_swipe_container.isRefreshing = true
                refreshList()
                list_swipe_container.isRefreshing = false
            }
            list_swipe_container.isRefreshing = false
        }

        list_swipe_container.setColorSchemeResources(
            android.R.color.holo_blue_bright,
            android.R.color.holo_green_light,
            android.R.color.holo_orange_light,
            android.R.color.holo_red_light
        )

        searchView.apply {
            setOnQueryTextListener(object : OnQueryTextListener,
                android.widget.SearchView.OnQueryTextListener {
                override fun onQueryTextSubmit(query: String?): Boolean {
                    isSearching = true
                    query?.let { photoViewModel.setQuery(it) }
                    photoViewModel.list.removeObservers(this@ListFragment)
                    query?.let { photoViewModel.getList(it) }
                    photoViewModel.list.observe(viewLifecycleOwner, observer)
                    return false
                }

                override fun onQueryTextChange(newText: String?): Boolean {
                    return false
                }

            })

            setOnCloseListener {
                if (isSearching) {
                    isSearching = false
                    searchView.onActionViewCollapsed()
                    photoViewModel.list.removeObservers(this@ListFragment)
                    photoViewModel.getList("")
                    photoViewModel.list.observe(viewLifecycleOwner, observer)
                    refreshList()
                    true
                } else {
                    searchView.onActionViewCollapsed()
                    true
                }
            }
        }

        morePopup.setOnClickListener {
            val popup = PopupMenu(requireContext(), it)
            popup.apply {
                menuInflater.inflate(R.menu.more_menu_popup, popup.menu)
                setOnMenuItemClickListener { item ->
                    when (item.itemId) {
                        R.id.item_logout -> {
                            val intent = Intent(context, MainActivity::class.java)
                            intent.addFlags(FLAG_ACTIVITY_NEW_TASK)
                            requireContext().startActivity(intent)
                            if (context is Activity) {
                                PreferenceManager.getDefaultSharedPreferences((activity as MainActivity).application)
                                    .edit().clear().apply()
                                (context as Activity).finish()
                            }
                            true
                        }
                        else -> true
                    }
                }
                show()
            }
        }
    }

    private fun photoClick(view: View, photo: Photo, position: Int) {
        val bundle = Bundle()
        bundle.apply {
            putString("RAW", photo.urls?.raw)
            putString("URI", photo.urls?.regular)
            photo.likes?.let { putInt("SMTH", it) }
            putString("ID", photo.id)
            photo.likedByUser?.let { putBoolean("ISLIKED", it) }
            putInt("POS", position)
        }

        val imageFragment = ImageFragment()
        imageFragment.arguments = bundle

        activity?.supportFragmentManager?.beginTransaction()
            ?.setReorderingAllowed(true)
            ?.replace(R.id.container, imageFragment)
            ?.addToBackStack("image")
            ?.addSharedElement(view, view.transitionName)
            ?.commit()
    }
}
